### PR TITLE
Influxdb Hook

### LINKED ISSUE #14168: add InfluxdbHook
**Description**
https://www.influxdata.com/
Add the ability to query and modify Influxdb from airflow.

Python wrapper that can be used to create the hook :
https://github.com/influxdata/influxdb-python

There are some usage examples in https://github.com/influxdata/influxdb-python/tree/master/examples
would be nice if the hook would support the following:
create_database
query
drop_database
write_points 

**Are you willing to submit a PR?**

not at the moment due to time contains.

**Related Issues**

couldn't find any

### PR DESCRIPTION
**Pull request by subkanthi**
**^ Added influxdb hook to perform operations on InfluxDB 2 using python library **
closes: #14168

---
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by uranusjr — CHANGES_REQUESTED** [review]
Issues from a quick scan:

* Do you intend to add content to `README`? If not, please remove the empty file.
* Please use the recommended module naming convention. Specifically, modules should use `lower_case_snake`.
* Casing is inconsistent; should it be `InfluxDB` or `Influxdb`? (And I believe there’s a casing mismatch in `provider.yml` that would cause tests to fail.)

**Comment by subkanthi** [comment]
> * InfluxDB

Thanks @uranusjr , Removed README and fixed the influxdb case in provider.yaml

**Review by potiuk — APPROVED** [review]
Nice! Good job! @uranusjr - any more comments? I am (finally) getting to release new providers and I hope I can get that one in!

**Comment by potiuk** [comment]
@uranusjr ?:)

File: airflow/providers/influxdb/hooks/influxdb.py
Comment (uranusjr) [inline airflow/providers/influxdb/hooks/influxdb.py]:
```suggestion
    def query_to_df(self, query) -> pd.DataFrame:
```

“With” to me means `query` is a dataframe, not the return value.

(`query_df` or `query_data_frame` like how the Influx DB client calls it is fine as well.)

**Review by uranusjr — APPROVED** [review]

File: airflow/providers/influxdb/example_dags/example_influxdb.py
Comment (uranusjr) [inline airflow/providers/influxdb/example_dags/example_influxdb.py:24]: Don’t name things with `test_`, this will be picked up by the test runner.

File: airflow/providers/influxdb/hooks/influxdb.py
Comment (uranusjr) [inline airflow/providers/influxdb/hooks/influxdb.py:67]: Is this `f""` needed? I think `conn.schema` is always a string here.

File: tests/providers/influxdb/operators/test_influxDbHook.py
Comment (uranusjr) [inline tests/providers/influxdb/operators/test_influxDbHook.py]: Would be nice to assert some calls on these mocks as well.

File: airflow/providers/influxdb/example_dags/example_influxdb.py
Comment (subkanthi) [reply airflow/providers/influxdb/example_dags/example_influxdb.py:24]: Fixed.

File: airflow/providers/influxdb/hooks/influxdb.py
Comment (subkanthi) [reply airflow/providers/influxdb/hooks/influxdb.py:67]: Removed f for conn.schema

File: tests/providers/influxdb/operators/test_influxDbHook.py
Comment (subkanthi) [reply tests/providers/influxdb/operators/test_influxDbHook.py]: Added asserts. Thanks

**Review by uranusjr — APPROVED** [review]
LGTM now.

**Review by potiuk — APPROVED** [review]
LGTM and heppy to merge as soon as as you rebase @subkanthi  - there were some unrelated errors in main (fixed since then). Can you please rebase?

**Comment by subkanthi** [comment]
> LGTM and heppy to merge as soon as as you rebase @subkanthi - there were some unrelated errors in main (fixed since then). Can you please rebase?

Thanks @potiuk for looking at it again, rebased.

File: airflow/providers/influxdb/provider.yaml
Comment (eladkal) [inline airflow/providers/influxdb/provider.yaml]: You have a test failure related to these lines:
```
  =================================== FAILURES ===================================
  _____________________ TestProviderManager.test_hook_values _____________________
  
  self = <test_providers_manager.TestProviderManager testMethod=test_hook_values>
  
      def test_hook_values(self):
          with pytest.warns(expected_warning=None) as warning_records:
              with self._caplog.at_level(logging.WARNING):
                  provider_manager = ProvidersManager()
                  connections_list = list(provider_manager.hooks.values())
                  assert len(connections_list) > 60
  >       assert [] == [w.message for w in warning_records.list if "hook-class-names" in str(w.message)]
  E       AssertionError: assert [] == [DeprecationW...rflow < 2.2")]
  E         Right contains one more item: DeprecationWarning("The provider apache-airflow-providers-influxdb uses `hook-class-names` property in provider-info a...f 'connection-types' in Airflow 2.2. Use **both** in case you want to have backwards compatibility with Airflow < 2.2")
  E         Use -v to get the full diff
  
  tests/core/test_providers_manager.py:132: AssertionError
```

You need to add also `connection-types` see https://github.com/apache/airflow/pull/17682

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/influxdb/provider.yaml
Comment (potiuk) [reply airflow/providers/influxdb/provider.yaml]: Glad that I added the test then :).

**Comment by eladkal** [comment]
@subkanthi can you rebase? there are failures on `test_current_user_has_permissions` which were fixed on main

**Comment by subkanthi** [comment]
> @subkanthi can you rebase? there are failures on `test_current_user_has_permissions` which were fixed on main

@eladkal Thanks rebased from master

File: airflow/providers/influxdb/example_dags/example_influxdb.py
Comment (eladkal) [inline airflow/providers/influxdb/example_dags/example_influxdb.py:1]: please add `__init__` to `example_dags`

**Review by eladkal — CHANGES_REQUESTED** [review]
There seems to still be issues with rebase. The errors shown on `tableau`
https://github.com/apache/airflow/pull/17068/checks?check_run_id=3557156642#step:6:5710
which doesn't seem related.

lets try another rebase and if it's persisted we will get to the bottom of this

File: airflow/providers/influxdb/example_dags/example_influxdb.py
Comment (subkanthi) [reply airflow/providers/influxdb/example_dags/example_influxdb.py:1]: Thanks @eladkal , added and rebased again, fingers crossed.

**Comment by potiuk** [comment]
There is indeed something fishy going on here with test collection. I am taking a look, 

BTW (for the future @eladkal @subkanthi ) - if you look at the logs, it is super easy to reproduce the very exact image/environment that was used during the build - in such cases it is really helpful.

```
 ./breeze --github-image-id 63d117d2fed6c0e9bfcfb09468fae347a944f03b --backend mssql --python 3.7 --db-reset --skip-mounting-local-sources --test-type Providers shell
```

**Comment by potiuk** [comment]
also @uranusjr -> for the future ^^ - for similar investigations, you might not be aware of it but we have full reproducibility of the images used during tests locally (the command is printed in the logs)

**Comment by potiuk** [comment]
Found the problem. Fix is coming. I am not sure how it was triggered now but was not triggered before. Tableau test package misses ``_init_.py` simply ... I will add it together with some protection to check for `__init__.py` also in `tests` - we already have it for "airflow.providers" - but we do not have it for `tests.providers` :(

**Comment by potiuk** [comment]
Fix here: https://github.com/apache/airflow/pull/18142 (there were quite a few missing ``__init__.py` files. Stil do not what triggered the error though (and did not trigger it before)

**Comment by potiuk** [comment]
Sorry for that @subkanthi ... That was somewhat unexpected :(

**Comment by potiuk** [comment]
OK. You can rebase (again :( @subkanthi )  - the fix to missing ``__init__.py` is merged.

**Comment by subkanthi** [comment]
> Sorry for that @subkanthi ... That was somewhat unexpected :(

No worries, thanks for helping out.

**Comment by potiuk** [comment]
One more missing ``__init__.py`` - now I  think I know where it could come from. Lack of __init__.py creates an implicit namespace and the problem was that from influxdb tests (in their own namespace) through some chain of imports the tests tried to load the "tableau"  and then they were in two separate namespaces and could not see each other.

**Comment by potiuk** [comment]
The only 'real' failure is build docs now :)

File: docs/apache-airflow-providers-influxdb/operators/influxdb.rst
Comment (eladkal) [inline docs/apache-airflow-providers-influxdb/operators/influxdb.rst:23]: Why do we need this if not implemented? I assume you added it because of the how-to-guide in the `provider.yaml` ? you can remove the `how-to-guide` from the yaml as there is no content here.

File: tests/providers/influxdb/operators/test_influxDbHook.py
Comment (eladkal) [inline tests/providers/influxdb/operators/test_influxDbHook.py]: This file name and path is wrong. This PR introduce only hook so it should not be in `/operators`
The path should be:
`tests/providers/influxdb/hooks/test_influxdb.py` (note you will also need to change the `__init__`)

also you only tested the connection. the  hook contains functions `query`/`create_organization`/`find_bucket_id_by_name` and others that have no testing. is this intentional?

**Review by eladkal — COMMENTED** [review]
Overall LGTM.
Left some comments

File: docs/apache-airflow-providers-influxdb/operators/influxdb.rst
Comment (subkanthi) [reply docs/apache-airflow-providers-influxdb/operators/influxdb.rst:23]: There was a build docs error, if this wasnt added. Let me try again, the problem is somehow Im not able to reproduce doc errors that happen in the CI server locally, but its got better

File: tests/providers/influxdb/operators/test_influxDbHook.py
Comment (subkanthi) [reply tests/providers/influxdb/operators/test_influxDbHook.py]: Thanks, will move it to hooks, some of the functions are simple wrappers, but I can add tests for two functions(write and find_bucket_id_by_name)

File: tests/providers/influxdb/hooks/test_influxDbHook.py
Comment (eladkal) [inline tests/providers/influxdb/hooks/test_influxDbHook.py]: You missed the rename of the file from `test_influxDBHook.py` to `test_influxdb.py`

Comment (subkanthi) [reply tests/providers/influxdb/hooks/test_influxDbHook.py]: Fixed and added more tests, @eladkal , thanks

**Review by eladkal — APPROVED** [review]
LGTM
I'll merge when CI is green

**Comment by potiuk** [comment]
Close/Reopen to rebuild

**Comment by rsingh2411** [comment]
@potiuk  any timelines for this feature , which release will be it part of.

**Comment by potiuk** [comment]
It's gone come in September wave of Providers ~ 1.5 week or so

**Comment by potiuk** [comment]
🤞

**Comment by subkanthi** [comment]
> 🤞

It looks much better than before, failing ones seem to azure related.

**Comment by eladkal** [comment]
> It looks much better than before, failing ones seem to azure related.

It's probably due to https://github.com/Azure/azure-sdk-for-python/pull/18801

Static failures needs to be addressed. we probably need to limit `azure-storage-blob` to < 12.9.0 in a separate PR

File: airflow/providers/influxdb/hooks/influxdb.py
Comment (eladkal) [inline airflow/providers/influxdb/hooks/influxdb.py:23]: mmm this provider must have `pandas` yet pandas is now an optional dependency in Airflow core https://github.com/apache/airflow/pull/17575
shouldn't we wrap it with try catch and raising exception if pandas is not installed?
@potiuk WDYT?

Comment (potiuk) [reply airflow/providers/influxdb/hooks/influxdb.py:23]: Good catch. I think It's better to get influx db extra to depend on pandas directly (with 'pandas>=0.17.1, <2.0`  - Ok with that @kanthi?

**Comment by potiuk** [comment]
> > It looks much better than before, failing ones seem to azure related.
> 
> It's probably due to [Azure/azure-sdk-for-python#18801](https://github.com/Azure/azure-sdk-for-python/pull/18801)
> 
> Static failures needs to be addressed. we probably need to limit `azure-storage-blob` to < 12.9.0 in a separate PR

Yep. That's what we discussed in slack. I will make that PR.

**Comment by potiuk** [comment]
https://github.com/apache/airflow/pull/18443 @subkanthi -  once this is merged you can rebase with additional pandas deps and 🤞 it will be good to merge.

**Comment by potiuk** [comment]
OK. I merged the change for azure, Just rebase now + add pandas to influx extra.

**Comment by subkanthi** [comment]
> OK. I merged the change for azure, Just rebase now + add pandas to influx extra.

Thanks @potiuk , will do now

File: airflow/providers/influxdb/hooks/influxdb.py
Comment (subkanthi) [reply airflow/providers/influxdb/hooks/influxdb.py:23]: Thanks, added in setup.py

**Comment by potiuk** [comment]
Hey @subkanthi - one LAST rebase please. I think it should be good to go - maybe you can also squash all commits ?

**Comment by subkanthi** [comment]
> Hey @subkanthi - one LAST rebase please. I think it should be good to go - maybe you can also squash all commits ?

Hi @potiuk , did a rebase, please let me know if its fine.

**Comment by potiuk** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

File: docs/apache-airflow-providers-influxdb/operators/influxdb.rst
Comment (eladkal) [reply docs/apache-airflow-providers-influxdb/operators/influxdb.rst:23]: @subkanthi Now that this is merged I think it a nice followup task would be to add `InfluxDBOperator`  around `query` function of the hook. It seems very simple and probably useful it will also help to add some content into this rst :)

**Pull request merged**
